{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Loading a dataset with ANUBIS (optional)\n",
    "\n",
    "Although not required for this analysis, ANUBIS package provides additional functions to work with insertion data. If needed, you can install this python package from [ANUBIS](https://github.com/CRG-CNAG/anubis) and then run the following code to obtain an anubis dataset:\n",
    "\n",
    "```python\n",
    "import anubis_core as ac\n",
    "all_reads = {}\n",
    "allfiles = {your_ide1:path1.qins, ..., your_ideN:pathN.qins}\n",
    "\n",
    "for k, v in allfiles.items():\n",
    "    all_reads[k] = qt.dataset(v, genome_location).zreads\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Case example with *Mycoplasma pneumoniae*\n",
    "\n",
    "Here we present a series of analysis used for the associated publication. We start first loading the required packages and configurations for this notebook:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Jupyter magics\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib widget\n",
    "\n",
    "#%matplotlib inline\n",
    "\n",
    "import re\n",
    "import glob\n",
    "import pickle\n",
    "import sys, os\n",
    "import utils as u\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import Counter\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "sns.set_style(\"whitegrid\", {'axes.grid' : False})\n",
    "\n",
    "from IPython.display import clear_output\n",
    "clear_output(wait=True)\n",
    "\n",
    "# Custom functions from protinseq_tools.py (provided along this notebook)\n",
    "import protinseq_tools as pst"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.1. Comprehensive omics information for *M. pneumoniae*\n",
    "\n",
    "We also provide a series of files including relevant information in `./data/` for plotting (such as associate omic data) and a minimal control-condition library example (Cm A and Cm B grown with 15 ug/ml of chloramphenicol) to showcase some analysis. This information is mostly saved in pickle format for which we provide a basic set of functions to upload. The files included are:\n",
    "\n",
    "- `example_reads.dat`: a dictionary {ide:array} where ide represents the unique identifier of the library and array includes the read count of insertion per genome base. \n",
    "- `myco_pneumoniae_m129.gb`: latest version of the genome of *M. pneumoniae* in genbank format.\n",
    "- `orfs_df.tsv`: a dataframe with the start, end and strand annotation of the ORFs to explore (you can produce this with RanSEPs orffinder function).\n",
    "- `annotations.dat`: a list of dictionaries with orfs, ncbi genes, functional rnas and MPN to mpneu translation dictionaries. \n",
    "- `intergenic.dat`: a dictionary with intergenic annotations. \n",
    "- `sequences.dat`: a list of dictionaries with nt and aa sequences. \n",
    "- `omics.dat`: RNA-Seq pile-ups and peptides identified by MS. \n",
    "- `mycoinfo.tsv`: a dataframe containing comprehensive information for each annotation analyzed. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/smiravet/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3063: DtypeWarning: Columns (8,9,22,29) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "all_reads                              = pst.LoadPickle('./data/example_reads.dat')\n",
    "genome_location                        ='./data/myco_pneumoniae_m129.gb'            # We provide just the location\n",
    "orfs_df                                = pd.read_csv('./data/orfs_df.tsv', sep='\\t', index_col=0)\n",
    "orfs, ncbi, rnas, mpn2mpneu, mpneu2mpn = pst.LoadPickle('./data/annotations.dat')\n",
    "igan                                   = pst.LoadPickle('./data/intergenic.dat')\n",
    "nt_seqs, aa_seqs                       = pst.LoadPickle('./data/sequences.dat')\n",
    "piles, peptide_map                     = pst.LoadPickle('./data/omics.dat')\n",
    "extmycoinfo                            = pd.read_csv('./data/mycoinfo.tsv', sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Cm_A_15_neg_1': array([3., 0., 0., ..., 0., 0., 0.]),\n",
       " 'Cm_A_15_pos_1': array([0., 0., 0., ..., 0., 0., 0.]),\n",
       " 'Cm_A_15_neg_3': array([0., 0., 0., ..., 0., 0., 0.]),\n",
       " 'Cm_A_15_pos_3': array([0., 0., 0., ..., 0., 0., 0.]),\n",
       " 'Cm_B_15_pos_3': array([0., 0., 0., ..., 0., 0., 0.]),\n",
       " 'Cm_B_15_neg_1': array([0., 0., 0., ..., 0., 0., 0.]),\n",
       " 'Cm_B_15_neg_3': array([0., 0., 0., ..., 0., 0., 0.]),\n",
       " 'Cm_B_15_pos_1': array([0., 0., 0., ..., 0., 0., 0.]),\n",
       " 'Cm_B_15_pos_2': array([0., 0., 0., ..., 0., 0., 0.]),\n",
       " 'Cm_B_15_neg_2': array([0., 0., 0., ..., 0., 0., 0.])}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>alt_ann</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>mpneu04917</th>\n",
       "      <td>38</td>\n",
       "      <td>58</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mpneu30113</th>\n",
       "      <td>33</td>\n",
       "      <td>47</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mpneu00001</th>\n",
       "      <td>49</td>\n",
       "      <td>132</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mpneu09814</th>\n",
       "      <td>60</td>\n",
       "      <td>173</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mpneu04918</th>\n",
       "      <td>62</td>\n",
       "      <td>94</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mpneu14813</th>\n",
       "      <td>816260</td>\n",
       "      <td>816277</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mpneu09813</th>\n",
       "      <td>816293</td>\n",
       "      <td>816322</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mpneu14812</th>\n",
       "      <td>816299</td>\n",
       "      <td>816325</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mpneu25196</th>\n",
       "      <td>815526</td>\n",
       "      <td>816338</td>\n",
       "      <td>-</td>\n",
       "      <td>MPN688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mpneu14811</th>\n",
       "      <td>816339</td>\n",
       "      <td>816362</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30112 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             start     end strand alt_ann\n",
       "gene                                     \n",
       "mpneu04917      38      58      +      -1\n",
       "mpneu30113      33      47      -      -1\n",
       "mpneu00001      49     132      +       0\n",
       "mpneu09814      60     173      +       0\n",
       "mpneu04918      62      94      +      -1\n",
       "...            ...     ...    ...     ...\n",
       "mpneu14813  816260  816277      -      -1\n",
       "mpneu09813  816293  816322      +      -1\n",
       "mpneu14812  816299  816325      -      -1\n",
       "mpneu25196  815526  816338      -  MPN688\n",
       "mpneu14811  816339  816362      +      -1\n",
       "\n",
       "[30112 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orfs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ntlen</th>\n",
       "      <th>aalen</th>\n",
       "      <th>ann_type</th>\n",
       "      <th>alt_ann</th>\n",
       "      <th>localization</th>\n",
       "      <th>function</th>\n",
       "      <th>operon</th>\n",
       "      <th>expression</th>\n",
       "      <th>cprna</th>\n",
       "      <th>msexp</th>\n",
       "      <th>mspepts</th>\n",
       "      <th>cpprot</th>\n",
       "      <th>ureacp</th>\n",
       "      <th>sdscp</th>\n",
       "      <th>meancp</th>\n",
       "      <th>hlprot</th>\n",
       "      <th>refdens</th>\n",
       "      <th>refesse</th>\n",
       "      <th>ranseps</th>\n",
       "      <th>rbs</th>\n",
       "      <th>homology</th>\n",
       "      <th>predfun</th>\n",
       "      <th>tm</th>\n",
       "      <th>sp</th>\n",
       "      <th>segtm</th>\n",
       "      <th>complex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MPN001</td>\n",
       "      <td>692</td>\n",
       "      <td>1834</td>\n",
       "      <td>+</td>\n",
       "      <td>1142</td>\n",
       "      <td>380.0</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>1</td>\n",
       "      <td>Cytoplasm</td>\n",
       "      <td>DNA synthesis/modifications/repair/conformation</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.459389</td>\n",
       "      <td>0.226970</td>\n",
       "      <td>116.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>508.1</td>\n",
       "      <td>508.1</td>\n",
       "      <td>663.5</td>\n",
       "      <td>867.03</td>\n",
       "      <td>8.06</td>\n",
       "      <td>0.004</td>\n",
       "      <td>E</td>\n",
       "      <td>0.993</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>o</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MPN002</td>\n",
       "      <td>1838</td>\n",
       "      <td>2767</td>\n",
       "      <td>+</td>\n",
       "      <td>929</td>\n",
       "      <td>309.0</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>1</td>\n",
       "      <td>Cytoplasm</td>\n",
       "      <td>DNA synthesis/modifications/repair/conformation</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.297465</td>\n",
       "      <td>0.126930</td>\n",
       "      <td>115.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>34.9</td>\n",
       "      <td>34.9</td>\n",
       "      <td>36.2</td>\n",
       "      <td>66.62</td>\n",
       "      <td>4.83</td>\n",
       "      <td>0.208</td>\n",
       "      <td>F</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>xdj1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>o</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MPN003</td>\n",
       "      <td>2869</td>\n",
       "      <td>4821</td>\n",
       "      <td>+</td>\n",
       "      <td>1952</td>\n",
       "      <td>650.0</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>1</td>\n",
       "      <td>Cytoplasm</td>\n",
       "      <td>DNA synthesis/modifications/repair/conformation</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.597315</td>\n",
       "      <td>0.234930</td>\n",
       "      <td>115.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>267.8</td>\n",
       "      <td>267.8</td>\n",
       "      <td>350.9</td>\n",
       "      <td>268.72</td>\n",
       "      <td>6.07</td>\n",
       "      <td>0.003</td>\n",
       "      <td>E</td>\n",
       "      <td>0.996</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>o</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MPN004</td>\n",
       "      <td>4821</td>\n",
       "      <td>7340</td>\n",
       "      <td>+</td>\n",
       "      <td>2519</td>\n",
       "      <td>839.0</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>1</td>\n",
       "      <td>Cytoplasm</td>\n",
       "      <td>DNA synthesis/modifications/repair/conformation</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.297611</td>\n",
       "      <td>0.183630</td>\n",
       "      <td>116.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>280.6</td>\n",
       "      <td>280.6</td>\n",
       "      <td>359.0</td>\n",
       "      <td>244.59</td>\n",
       "      <td>7.05</td>\n",
       "      <td>0.002</td>\n",
       "      <td>E</td>\n",
       "      <td>0.994</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>o</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MPN005</td>\n",
       "      <td>7312</td>\n",
       "      <td>8574</td>\n",
       "      <td>+</td>\n",
       "      <td>1262</td>\n",
       "      <td>420.0</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>1</td>\n",
       "      <td>Cytoplasm</td>\n",
       "      <td>tRNA synthases</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.974636</td>\n",
       "      <td>0.177735</td>\n",
       "      <td>115.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>296.0</td>\n",
       "      <td>296.0</td>\n",
       "      <td>177.8</td>\n",
       "      <td>249.56</td>\n",
       "      <td>7.87</td>\n",
       "      <td>0.002</td>\n",
       "      <td>E</td>\n",
       "      <td>0.984</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>serS</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>o</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34196</th>\n",
       "      <td>IGMM1659</td>\n",
       "      <td>800442</td>\n",
       "      <td>800542</td>\n",
       "      <td>-</td>\n",
       "      <td>100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igan</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Functional RNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34197</th>\n",
       "      <td>IGMM1660</td>\n",
       "      <td>800627</td>\n",
       "      <td>800644</td>\n",
       "      <td>-</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igan</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Functional RNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34198</th>\n",
       "      <td>IGMM1661</td>\n",
       "      <td>805849</td>\n",
       "      <td>805870</td>\n",
       "      <td>-</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igan</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Functional RNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34199</th>\n",
       "      <td>IGMM1662</td>\n",
       "      <td>813449</td>\n",
       "      <td>813467</td>\n",
       "      <td>-</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igan</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Functional RNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34200</th>\n",
       "      <td>IGMM1663</td>\n",
       "      <td>816339</td>\n",
       "      <td>816394</td>\n",
       "      <td>-</td>\n",
       "      <td>55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igan</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Functional RNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34201 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene   start     end strand  ntlen  aalen ann_type alt_ann  \\\n",
       "0        MPN001     692    1834      +   1142  380.0     ncbi       1   \n",
       "1        MPN002    1838    2767      +    929  309.0     ncbi       1   \n",
       "2        MPN003    2869    4821      +   1952  650.0     ncbi       1   \n",
       "3        MPN004    4821    7340      +   2519  839.0     ncbi       1   \n",
       "4        MPN005    7312    8574      +   1262  420.0     ncbi       1   \n",
       "...         ...     ...     ...    ...    ...    ...      ...     ...   \n",
       "34196  IGMM1659  800442  800542      -    100    NaN     igan       0   \n",
       "34197  IGMM1660  800627  800644      -     17    NaN     igan       0   \n",
       "34198  IGMM1661  805849  805870      -     21    NaN     igan       0   \n",
       "34199  IGMM1662  813449  813467      -     18    NaN     igan       0   \n",
       "34200  IGMM1663  816339  816394      -     55    NaN     igan       0   \n",
       "\n",
       "      localization                                         function  operon  \\\n",
       "0        Cytoplasm  DNA synthesis/modifications/repair/conformation     1.0   \n",
       "1        Cytoplasm  DNA synthesis/modifications/repair/conformation     1.0   \n",
       "2        Cytoplasm  DNA synthesis/modifications/repair/conformation     2.0   \n",
       "3        Cytoplasm  DNA synthesis/modifications/repair/conformation     2.0   \n",
       "4        Cytoplasm                                   tRNA synthases     2.0   \n",
       "...            ...                                              ...     ...   \n",
       "34196          NaN                                   Functional RNA     NaN   \n",
       "34197          NaN                                   Functional RNA     NaN   \n",
       "34198          NaN                                   Functional RNA     NaN   \n",
       "34199          NaN                                   Functional RNA     NaN   \n",
       "34200          NaN                                   Functional RNA     NaN   \n",
       "\n",
       "       expression     cprna  msexp  mspepts  cpprot  ureacp  sdscp  meancp  \\\n",
       "0        8.459389  0.226970  116.0     33.0   508.1   508.1  663.5  867.03   \n",
       "1        7.297465  0.126930  115.0     22.0    34.9    34.9   36.2   66.62   \n",
       "2        8.597315  0.234930  115.0     48.0   267.8   267.8  350.9  268.72   \n",
       "3        8.297611  0.183630  116.0     54.0   280.6   280.6  359.0  244.59   \n",
       "4        7.974636  0.177735  115.0     28.0   296.0   296.0  177.8  249.56   \n",
       "...           ...       ...    ...      ...     ...     ...    ...     ...   \n",
       "34196         NaN       NaN    0.0      0.0     0.0     0.0    0.0    0.00   \n",
       "34197         NaN       NaN    0.0      0.0     0.0     0.0    0.0    0.00   \n",
       "34198         NaN       NaN    0.0      0.0     0.0     0.0    0.0    0.00   \n",
       "34199         NaN       NaN    0.0      0.0     0.0     0.0    0.0    0.00   \n",
       "34200         NaN       NaN    0.0      0.0     0.0     0.0    0.0    0.00   \n",
       "\n",
       "       hlprot  refdens refesse  ranseps  rbs  homology predfun   tm   sp  \\\n",
       "0        8.06    0.004       E    0.993  1.0       6.0    dnaN  0.0  0.0   \n",
       "1        4.83    0.208       F    0.998  0.0       6.0    xdj1  0.0  0.0   \n",
       "2        6.07    0.003       E    0.996  1.0       6.0    gyrB  0.0  0.0   \n",
       "3        7.05    0.002       E    0.994  0.0       6.0    gyrA  0.0  0.0   \n",
       "4        7.87    0.002       E    0.984  1.0       6.0    serS  0.0  0.0   \n",
       "...       ...      ...     ...      ...  ...       ...     ...  ...  ...   \n",
       "34196    0.00      NaN     NaN    0.000  0.0       0.0     RNA  0.0  0.0   \n",
       "34197    0.00      NaN     NaN    0.000  0.0       0.0     RNA  0.0  0.0   \n",
       "34198    0.00      NaN     NaN    0.000  0.0       0.0     RNA  0.0  0.0   \n",
       "34199    0.00      NaN     NaN    0.000  0.0       0.0     RNA  0.0  0.0   \n",
       "34200    0.00      NaN     NaN    0.000  0.0       0.0     RNA  0.0  0.0   \n",
       "\n",
       "      segtm  complex  \n",
       "0         o      2.0  \n",
       "1         o      1.0  \n",
       "2         o      2.0  \n",
       "3         o      2.0  \n",
       "4         o      2.0  \n",
       "...     ...      ...  \n",
       "34196     0      0.0  \n",
       "34197     0      0.0  \n",
       "34198     0      0.0  \n",
       "34199     0      0.0  \n",
       "34200     0      0.0  \n",
       "\n",
       "[34201 rows x 30 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "extmycoinfo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2. Metrics by frame\n",
    "\n",
    "To retrieve the main metrics by frame for an annotation in the library:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>frame</th>\n",
       "      <th>metric</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>R</td>\n",
       "      <td>47487.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>407.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>coverage</td>\n",
       "      <td>0.399542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>R</td>\n",
       "      <td>360.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>I</td>\n",
       "      <td>24.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>coverage</td>\n",
       "      <td>0.023560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3</td>\n",
       "      <td>R</td>\n",
       "      <td>30.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3</td>\n",
       "      <td>I</td>\n",
       "      <td>13.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3</td>\n",
       "      <td>coverage</td>\n",
       "      <td>0.012762</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   frame    metric         value\n",
       "0      1         R  47487.000000\n",
       "1      1         I    407.000000\n",
       "2      1  coverage      0.399542\n",
       "3      2         R    360.000000\n",
       "4      2         I     24.000000\n",
       "5      2  coverage      0.023560\n",
       "6      3         R     30.000000\n",
       "7      3         I     13.000000\n",
       "8      3  coverage      0.012762"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pst.get_f_metric(all_reads['Cm_B_15_neg_1'], ncbi['MPN447'], thr=0) # Threshold can be used to remove insertions by minimum value of reads. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bb414a0975b048389edbb0d3fd929fcd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'Nr Ins')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pst.get_f_metric(all_reads['Cm_B_15_neg_1'], ncbi['MPN447'])\n",
    "plt.figure()\n",
    "sns.barplot(x='frame', y='value', data=df[df['metric']=='I'])\n",
    "plt.ylabel('Nr Ins')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It can be observThis can be iterated to collect multiple metrics per sample for genes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3. Basic plotting function\n",
    "\n",
    "This function interacts with the function pst.psplot function to provide complex visualizations of the insertions profiles:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_ide(ide='region', coords=None,\n",
    "             query=None, wdw_size=5, tracks=[1,1,1,1,1,1,1], strand_specific=True, extra=100, \n",
    "             list_of_samples=None, profile=None, alt_start=True):\n",
    "    if ide!='region':\n",
    "        coords = list(extmycoinfo.loc[extmycoinfo['gene']==ide, ['start', 'end', 'strand']].values[0])\n",
    "    elif coords:\n",
    "        alt_start=False\n",
    "        pass\n",
    "    if strand_specific:\n",
    "        if coords[-1]=='+':        \n",
    "            q = 'pos'\n",
    "        else:\n",
    "            q = 'neg'\n",
    "    # Define query\n",
    "    if profile:\n",
    "        profile = {'custom':profile}\n",
    "    else:\n",
    "        if list_of_samples:\n",
    "            profile = {k:all_reads[k] for k in list_of_samples}\n",
    "        elif query:\n",
    "            if type(query)==str:\n",
    "                query = [query]\n",
    "            if type(query[0])==str:\n",
    "                if strand_specific:\n",
    "                    query.append(q)\n",
    "                profile = ps.return_sample(all_reads, select_all=query)\n",
    "            elif type(query[0])==list:\n",
    "                if strand_specific:\n",
    "                    query[0].append(q)\n",
    "                if len(query)==2:\n",
    "                    profile = ps.return_sample(all_reads, select_all=query[0], select_any=query[1])\n",
    "                elif len(query)==3:\n",
    "                    profile = ps.return_sample(all_reads, select_all=query[0], select_any=query[1], exclude=query[2])\n",
    "            else:\n",
    "                sys.exit('provide accepted query')\n",
    "        else:\n",
    "            profile = None    \n",
    "    pst.psplot(coords=coords, ide=ide, \n",
    "               piles=piles, profile=profile, \n",
    "               nt_seq=nt_seqs.get(ide, None), alt_start=alt_start,\n",
    "               # query=query,\n",
    "               peptides=peptide_map.get(ide, None), tracks=tracks, wdw_size=5,\n",
    "               aa_seq=aa_seqs.get(ide, None), extra=extra, annotation=orfs_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3.1. Plotting a gene by name:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "59cbecf55d6b4b418c0f49cebf93d854",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_ide(mpn2mpneu['MPN001'], profile=list(all_reads['Cm_A_15_pos_1']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each row presents (in order):\n",
    "- RNA-Seq profile\n",
    "- unique tryptic peptides identified by Mass-spec\n",
    "- mapped insertions colored by frame (as many samples as provided)\n",
    "- the transmembrane prediction with TMHMM\n",
    "- the accumulated windows analysis of insertion profiles\n",
    "- the map of annotation in the region\n",
    "\n",
    "Notice the sample orientation identifier (positive strand) matches the one of the gene. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3.2. Plotting multiple profiles:\n",
    "\n",
    "The function will check directly in the all_reads dictionary, this can be use to plot all replicates with a given text match. In this case, we plot the gene MPN593. Green lines in the transmembrane track represent 'ruptures' in the profile, these are points of changes in the distribution for the insertion prediction which in this case are representative of the transmembrane protein topology:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fe79d6a60aad40718b3f961f3de023d0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_ide(ide='mpneu25694', list_of_samples=[k for k in all_reads.keys() if 'Cm_B_15_neg' in k])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3.3. Selecting which elements to plot by coordinates:\n",
    "\n",
    "You can also plot specific coordinates hiding elements which are relative to a gene with the tracks argument. In this case we plot the example from the manuscript supplementary figure 7:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2b5ca1f2d5cd496181240b2fa2f19cb5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_ide(coords=[543300,553520, '-'], list_of_samples=[k for k in all_reads.keys() if '_15_neg_1' in k], tracks=[0,0,1,0,0,0,0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.4. Metagene comparative\n",
    "\n",
    "A metagene can be quickly retrieved for a list of annotations by:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "your_annotations = [v for k,v in ncbi.items() if v[-1]=='-']   # annotations from NCBI in the complementary strand\n",
    "metagene = pst.metagene_by_frame(all_reads['Cm_B_15_neg_1'], your_annotations, frame=1, metric='R', step=100, gl=816394)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- accepted_metrics by the function include: `'sum', 'std', 'mean', 'median', 'count', 'L', 'min', 'max', 'I', 'dens', 'R', 'RI', 'CPM', 'RPKM'.`\n",
    "- step defines the number of bins (100 in this case). \n",
    "- gl represents the genome length, used for normalization purposes. \n",
    "\n",
    "Then this metagene can be plotted:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "98c6ffab85634552b82225455b7c52fd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize=(10,2))\n",
    "plt.plot(np.sum(metagene, axis=1))\n",
    "plt.ylabel('Sum of reads')\n",
    "plt.xlabel('bin')\n",
    "plt.tight_layout()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
